synapseclient/core/retry.py

```diff
@@ -340,7 +374,130 @@ async def foo(a, b, c): return [a, b, c]
                 (
                     "Retries have run out. re-raising the exception: %s"
                     if retry
-                    else "Rasing the exception: %s"
+                    else "Raising the exception: %s"
+                ),
+                str(caught_exception_info[0]),
+            )
+            raise caught_exception
+        return response
+
+
+def with_retry_time_based(
+    function,
```

Comment:
Nit: Have you looked at the backoff package?

Reply:
I had not - Reading over the docs it does handle for most if not all the use-cases here. I won't make changes since this is working, but it may be a candidate to swap to if this does not work out.

Reply:
This might be a good future project!